KEP-3070: Reserve Service IP Ranges For Dynamic and Static IP Allocation (#3071)
kubernetes/enhancements: keps/sig-network/3070-reserved-service-ip-range/README.md

           │   static    │                    dynamic                  │
           └─────────────┴─────────────────────────────────────────────┘

           ◄────────────► ◄────────────────────────────────────────────►
    192.168.0.1     192.168.0.16                                 192.168.0.62

```go
// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
// forward looking for the next available address. The big.Int range is subdivided so it will try
// to allocate first from the reserved block of addresses (it will wrap the reserved subrange if necessary).
// If there is no free address it will try to allocate one from the reserved block too.
type randomScanReservedStrategy struct {
	rand     *rand.Rand
	reserved int
}
```

### Test Plan

This feature doesn't modify the cluster behavior, only the order on which dynamic IP are assigned to Services,
there is no need for e2e or integration tests, unit tests are enough.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag

#### Beta

- Gather feedback from developers and community.
- No issues reported.

#### GA

> - 2 examples of real-world usage

Reviewer: What does this mean?  That 2 providers have switched to this mode as the default?  That's unlikely while it is beta.  Let's get specific here.  I think time and lack of trouble is the only real signal.

When do we make this mode the default?  Is there a reason not to do that?

Author: It is interest in enabling it in Openshift by default to solve some races at installation time, I was expecting more installers will face similar issue, but as you say this is just more a matter of time and signal.

### Upgrade / Downgrade Strategy

The feature only changes the allocation strategy, the bitmaps and the underlay logic remain the same,
guaranteeing full compatibility and no risk rolling back the feature.

### Version Skew Strategy

Same as with upgrade, there is no change in the bitmaps, keeping the feature 100% compatible.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ServiceIPStaticSubrange
  - Components depending on the feature gate: kube-apiserver
- [x] Other
  - Describe the mechanism:
  - The feature require to restart the apiserver to use the new allocation strategy for Services, however,
  it will not require downtime in multi-control-plane environments.
    
###### Does enabling the feature change any default behavior?

> Dynamic allocated IPs will not be chosen from the first 32 values of the Service IP range, that is transparent to the user,

Reviewer: This still says 32

since the IP is and was already random, it doesn't matter from within the range the IP is obtained.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes.

###### What happens if we reenable the feature if it was previously rolled back?

Nothing, this only affects new requests and how these new request obtain the dynamically assigned IPs.

###### Are there any tests for feature enablement/disablement?

No need to add new test, current tests should keep working as today.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Current Services will not be affected, it doesn't matter in which range their IPs are allocated.
This only affects the creation of new Services without an IP specified, there is no risk of impact on running workloads.

###### What specific metrics should inform a rollback?

> The allocation logic already has the following metrics:

Reviewer: Do we want new metrics to indicate "forced to use reserved range" ?

Author: I've tried to add new metrics but its complicated, since this behavior is deep in the bitmap allocation and the metrics are one layer above on the ip allocator .... inheritance problems :shrug: 

I'll add a comment explaining it

Reviewer 2: > I've tried to add new metrics but its complicated, since this behavior is deep in the bitmap allocation and the metrics are one layer above on the ip allocator .... inheritance problems

Hmm - can't we report them somewhat post-factum (when we're returning the actual result we already know what IP we allocated and if it was statically requested or not)?

Author: do you mean something like `randomly_allocated_ips` gauge and `randomly_allocation_total` counter ?
that is easy to do, will that work for you?

Reviewer 2: > do you mean something like randomly_allocated_ips gauge and randomly_allocation_total counter ?
that is easy to do, will that work for you?

Can you clarify? I think these can be interpreted differently.

Author: we have these metrics now
```
  - allocated_ips
  - available_ips
  - allocation_total
  - allocation_errors_total
```
we can add two new ones
```
randomly_allocated_ips
randomly_allocation_total
```
`randomly_allocated_ips` reports the number of IPs in the allocator that are allocated "randomly" i.e. with ClusterIP empty, it reports current value.
`randomly_allocation_total` is a counter, every time that you allocate a ClusterIP randomly it increases, the user can play with the current metrics and these new ones to obtain the IPs allocated statically

Reviewer 2: Thanks for clarifiction.

Yes - those metric will be very useful.

What I had on my mind (please let me know if this doesn't make any sense because of some reason) is actually an extension of that by:
- adding a label to that metric (like "from-static" or sth) to show if those random IPs were allocated from static range

Does that make sense?

Author: yeah, that is much simpler and easier, I will update the kep with that

  - allocated_ips
  - available_ips
  - allocation_total
  - allocation_errors_total

The increase of the errors metrics or the trend of the allocated_ips and available_ips doesn't change and new services are created or deleted.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

### Monitoring Requirements

Current allocator metrics are enough to understand the health of the allocation logic.

  - allocated_ips
  - available_ips
  - allocation_total
  - allocation_errors_total

> ###### How can an operator determine if the feature is in use by workloads?

Reviewer: I am embarrassed to not know this, but do we have a simple "settings" URL that lists things like this?

Author: Depending on the distro you can find the setting in different places, some use CRs to store the config, other store it in configmap , .... I don't know if there is some endpoint in the apiserver that gives you that, @liggitt may know better.

Reviewer: I feel like we should have something, akin to metrics, which publishes static (or nearly static) settings, with a simple api like `settingz.AddInt("service-ip-static-subrange-size", 16)`

Author: we can publish an info metrics with the allocator information, same that apiserver does with the build info 

https://github.com/kubernetes/kubernetes/blob/55e1d2f9a73a0c78fc0c266236c3e31261635b90/staging/src/k8s.io/component-base/metrics/prometheus/version/metrics.go#L26-L33

WDYT?

Reviewer 3: there's no /configz endpoint or equivalent. The things that crop up when this is requested for one-off things are:

1. who should have visibility to this
2. API structure and guarantees (e.g. should we communicate single configuration values or multiple? if multiple, in what format?)
3. behavior in multi-server clusters

Author: for the allocator I'll add the metrics with the constant  as in https://github.com/kubernetes/enhancements/pull/3071/files#r778133274 , the setting endpoint thing deserves its own kep :)

> # KEP-3070: Reserve Service IP Range For Dynamic and Static IP Allocation

Reviewer 4: either "Range for Static ..." or "Range**s** for Dynamic and Static"

Author: the later :+1:

- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Kubernetes Services are an abstract way to expose an application running on a
set of Pods. Services has a ClusterIP that is virtual and allows to load-balance
traffic across the different Pods. This ClusterIP can be assigned:

- dynamically, the cluster will pick one within the configured Service IP range.
- statically, the user will set one IP within the configured Service IP range.

Currently, there is no possibility, before creating a Service with a static Cluster IP,
to know if the IP has been chosen already by any other Service using dynamic allocation.

The Service IP range can be logically subdivided to avoid the risk of conflict between
Services that use static and dynamic IP allocation.

## Motivation

There are situation that users need to rely on well-known predefined IP addresses.

> The best example is the Kubernetes DNS, the IP is hardcoded in the kubelet flag `--cluster-dns`.

Reviewer 4: (Well, no, the _best_ example is the apiserver...)

Author: is it possible to race with the apiserver .1 address? 
The loop to create the `kubernetes.default` runs at the beginning, I honestly don't know if at that point clients are able to create services, but is worth adding it

           │   static    │                    dynamic                  │
           └─────────────┴─────────────────────────────────────────────┘

           ◄────────────► ◄────────────────────────────────────────────►
    192.168.0.1     192.168.0.64                                 192.168.3.254

Example 3 Small Subnet:

- Service IP CIDR: 192.168.0.0/26
- Range Size: 2^6 - 2 = 62
- Band Offset: min(max(16,64/16),256) = min(max(16,4),256) = 16
- Static band start: 192.168.0.1
- Static band ends: 192.168.0.16

           ┌─────────────┬─────────────────────────────────────────────┐
           │   static    │                    dynamic                  │
           └─────────────┴─────────────────────────────────────────────┘

           ◄────────────► ◄────────────────────────────────────────────►
    192.168.0.1     192.168.0.16                                 192.168.0.62

```go
// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
// forward looking for the next available address. The big.Int range is subdivided so it will try
// to allocate first from the reserved block of addresses (it will wrap the reserved subrange if necessary).

> // If there is no free address it will try to allocate one from the reserved block too.

Reviewer 4: "to allocate first from the **un**reserved block", right? (You're referring to both ranges as "the reserved block" here.)

and "(it will wrap around back to the start of the unreserved subrange if necessary)"

Author: I will word it as reserved for static and reserved for dynamic allocation, it sounds less ambiguous

	if count >= max {
		return 0, false
	}
	offset := rss.rand.Intn(max)
	for i := 0; i < max; i++ {
		at := (offset + i) % max
		if allocated.Bit(at) == 0 {
			return at, true
		}
	}
	return 0, false
}
```

### Proposed Services ClusterIPs allocation model

The proposal is to implement a new strategy on the Service IP allocation bitmap that logically subdivide the range
in two bands based on the following formula `min(max($min, cidrSize/$step), $max)`, described as ~never less than $min or more than $max, with a graduated step function between them~, with $min = 16, $max = 256 and $step = 16.

- lower band, used preferably for static ip assignment.
- upper band, used preferably for dynamic IP assignment.

Dynamically IP assignment will use the upper band by default, once this has been exhausted it will use the
lower range. This will allow users to define static allocations on the lower band of the predefined
range with a low risk of collision, keeping the implementation backwards compatible.

Example 1 Large Subnet:

- Service IP CIDR: 192.168.0.0/16
- Range Size: 2^16 - 2 = 65534
- Band Offset: min(max(16,65536/16),256) = min(4096,256) = 256
- Static band start: 192.168.0.1

> - Static band ends: 192.168.1.1

Reviewer 2: nit: shouldn't it be 192.168.1.0 ?

Author: off by one thing XD

```suggestion
- Static band ends: 192.168.1.0
```

           ┌─────────────┬─────────────────────────────────────────────┐
           │   static    │                    dynamic                  │
           └─────────────┴─────────────────────────────────────────────┘

           ◄────────────► ◄────────────────────────────────────────────►

>     192.168.0.1     192.168.1.1                                 192.168.255.254

Author: ```suggestion
    192.168.0.1     192.168.1.0                                 192.168.255.254
```

type randomScanReservedStrategy struct {
	rand     *rand.Rand
	reserved int
}
```

### Test Plan

This feature doesn't modify the cluster behavior, only the order on which dynamic IP are assigned to Services,
there is no need for e2e or integration tests, unit tests are enough.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag

#### Beta

- Gather feedback from developers and community.
- No issues reported.

#### GA

- No issues reported during two releases.

### Upgrade / Downgrade Strategy

The feature only changes the allocation strategy, the bitmaps and the underlay logic remain the same,
guaranteeing full compatibility and no risk rolling back the feature.

### Version Skew Strategy

Same as with upgrade, there is no change in the bitmaps, keeping the feature 100% compatible.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ServiceIPStaticSubrange
  - Components depending on the feature gate: kube-apiserver

> - [x] Other

Reviewer 2: Hmm - do we really need it?

We will have a feature gate (changing which requires restart of kube-apiserver anyway).
So can't we just initialize the proper allocation strategy based on feature-gate? [Then you don't need anything else...]

Author: I misunderstood this question

  - Describe the mechanism:
  - The feature require to restart the apiserver to use the new allocation strategy for Services, however,
  it will not require downtime in multi-control-plane environments.
    
###### Does enabling the feature change any default behavior?

> Dynamic allocated IPs will not be chosen from the first X values of the Service IP range, where X is obtained from the formula

Reviewer 2: nit: "initially chosen" maybe?

IIRC, if the second range is fully occupied, we will fallback to the reserved range, right?

Author: ```suggestion
Dynamic allocated IPs will not be initially chosen from the first X values of the Service IP range, where X is obtained from the formula
```

###### How can a rollout or rollback fail? Can it impact already running workloads?

Current Services will not be affected, it doesn't matter in which range their IPs are allocated.
This only affects the creation of new Services without an IP specified, there is no risk of impact on running workloads.

###### What specific metrics should inform a rollback?

The allocation logic already has the following metrics, that will be expanded with a new label containing information about the
type of allocation requested (dynamic or static):

  - allocated_ips
  - available_ips
  - allocation_total
  - allocation_errors_total

The increase of the errors metrics or the trend of the allocated_ips and available_ips doesn't change and new services are created or deleted.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

### Monitoring Requirements

Current allocator metrics will be expanded with a new label to each metric containing information about the
type of allocation requested (dynamic or static):

  - allocated_ips

>   - available_ips

Author: @wojtek-t these 2 metrics can not track the information of the type of allocation, because when an ip is released we don't have that information.
however, the counters can have it, because are only incremented
```
  - allocation_total
  - allocation_errors_total
```

Reviewer 2: Fair - that sounds good enough to me.

Can you clarify that here?

> Current allocator metrics will be expanded with a new label to each metric containing information about the
> type of allocation requested (dynamic or static):
> 
>   - allocated_ips
>   - available_ips
>   - allocation_total
>   - allocation_errors_total

Author: ```suggestion
Following allocator metrics will be expanded with a new label to each metric containing information about the
type of allocation requested (dynamic or static):

  - allocation_total
  - allocation_errors_total

The other allocator the metrics can not use the additional label because, when an ip is released, there is not information on how the ip was allocated.

    - allocated_ips
  - available_ips
```

###### How can an operator determine if the feature is in use by workloads?

A new metric will be added containing the information of the range configuration in the allocator.

```go
	allocatorInfo = metrics.NewGaugeVec(
		&metrics.GaugeOpts{

> 			Name:           "kubernetes_allocator_info",

Author: ```suggestion
			Name:           "allocator_info",
```

			Help:           "A metric with a constant '1' value labeled by Service CIDR, ",
			StabilityLevel: metrics.ALPHA,
		},

> 		[]string{"cidr", "address_min", "address_max", "dynamic_range_offset"},

Author: address min and max doesn't add much value, having the cidr you know what is the first address

I think better to use 
```suggestion
		[]string{"cidr", "allocator_size", "dynamic_range_offset"},
```

		etcd, err := serviceallocator.NewEtcd(mem, "/ranges/serviceips", serviceStorageConfig.ForResource(api.Resource("serviceipallocations")))
		if err != nil {
			return nil, err
		}
		serviceClusterIPRegistry = etcd
		return etcd, nil
	})
```

The bitmaps implement a [strategy for allocating a random free value if none is specified](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/pkg/registry/core/service/allocator/bitmap.go#L186-L205):

```go
// randomScanStrategy chooses a random address from the provided big.Int, and then
// scans forward looking for the next available address (it will wrap the range if
// necessary).
type randomScanStrategy struct {
	rand *rand.Rand
}

func (rss randomScanStrategy) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
	if count >= max {
		return 0, false
	}
	offset := rss.rand.Intn(max)
	for i := 0; i < max; i++ {
		at := (offset + i) % max
		if allocated.Bit(at) == 0 {
			return at, true
		}
	}
	return 0, false
}
```

### Proposed Services ClusterIPs allocation model

The proposal is to implement a new strategy on the Service IP allocation bitmap that logically subdivide the range

> in two bands based on the following formula `min(max($min, cidrSize/$step), $max)`, described as ~never less than $min or more than $max, with a graduated step function between them~, with $min = 16, $max = 256 and $step = 16.

Author: @wojtek-t @thockin there is a corner case, if the cidrsize is 16 or lower, the formula will not apply

Author: ```suggestion
The proposal is to implement a new strategy on the Service IP allocation bitmap that logically subdivide the range
in two bands based on the following formula `min(max($min, cidrSize/$step), $max)`, described as ~never less than $min or more than $max, with a graduated step function between them~, with $min = 16, $max = 256 and $step = 16, only if cidrSize < $min 
```

Reviewer 2: yeah - makes sense

<!--

> **Note:** When your KEP is complete, all of these comment blocks should be removed.

Reviewer 4: can you remove the blocks that are no longer relevant

marked as `provisional` is a working document and subject to change. You can
denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused. If you disagree with what is already in a document, open a new PR
with suggested changes.

One KEP corresponds to one "feature" or "enhancement" for its whole lifecycle.
You do not need a new KEP to move from beta to GA, for example. If
new details emerge that belong in the KEP, edit the KEP. Once a feature has become
"implemented", major changes should get new KEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/keps/NNNN-kep-template/README.md).

**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
it is marked `implementable`, must be approved by each of the KEP approvers.
If none of those approvers are still appropriate, then changes to that list
should be approved by the remaining approvers and/or the owning SIG (or
SIG Architecture for cross-cutting KEPs).
-->

> # KEP-3070: Reserve Service IP Range For Dynamic and Static IP Allocation

Reviewer 5: cna you add a sentence about your other IPAM kep w/ 600 comments, and how this is a "simpler" solution to part of that problem

Author: yeah, we'll add it to the alternatives section

[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Kubernetes Services are an abstract way to expose an application running on a
set of Pods. Services has a ClusterIP that is virtual and allows to load-balance
traffic accross the different Pods. This ClusterIP can be assigned:
- dynamically, the cluster will pick one within the configured Service IP range
- statically, the user will set one IP within the configured Service IP range

Currently, there is no possibility, when creating a Service with a static Cluster IP
to know if the IP has been chosen randomly by any other Service using dynamic allocation.

Defining a predefined range in the Service IP subnet to not allocate dynamic address from it,
allow to avoid any possible conflict with Services with static IP inside that range.

## Motivation

There are situation that users need to rely on well-known predefined IP addresses.

The best example is the Kubernetes DNS, the IP is hardcoded in the kubelet flag
`--cluster-dns strings`. For historical reasons and by convention the Service IP

> for the DNS use to the the 10th IP from the range, however, until the DNS Service

Reviewer 4: "the Service IP for DNS to use is the 10th IP from the range"?

has been created this IP can be taken by chance by any Service created with a dynamically
allocated IP.

There are also other use cases like the one explained in 
https://github.com/kubernetes/kubernetes/issues/95570

### Goals

- allow to create a Service with a static IPs without any risk of IP conflict.

### Non-Goals

- changing the allocators implementation (bitmaps, ...).
- add new configuration options.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

"As a developer I want to be able to create Services with well-known static IPs with the guarantee

> that those IPs will not be hijacked by a Service using dynamic IP allocation"

Reviewer 4: That's not a user story, that's a restatement of the Summary with the words "I want" before it.

The user stories here are (1) DNS getting the 10th IP so you can pass it to `--cluster-dns`, and (2) the use case from https://github.com/kubernetes/kubernetes/issues/95570

### Risks and Mitigations

There is no risk meanwhile every Service should be able to get an IP from the Service IP range
meanwhile there are free IPs.

This KEP only subdivides the Service IP range, and prioritize one range over other for dynamically
IP allocation, without adding limitations to the number of IPs assigned.

## Design Details

ClusterIP is the IP address of the service, if an address is specified manually, is in-range, and is not in use,

> it will be allocated to the service; otherwise creation of the service will fail.

Reviewer 4: If you're talking about the current/pre-KEP behavior here you should clarify that, since this section overall is for describing the proposed/post-KEP behavior.

		etcd, err := serviceallocator.NewEtcd(mem, "/ranges/serviceips", serviceStorageConfig.ForResource(api.Resource("serviceipallocations")))
		if err != nil {
			return nil, err
		}
		serviceClusterIPRegistry = etcd
		return etcd, nil
	})
```

The bitmaps implement a [strategy for allocating a random free value if none is specified](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/pkg/registry/core/service/allocator/bitmap.go#L186-L205):

```go
// randomScanStrategy chooses a random address from the provided big.Int, and then
// scans forward looking for the next available address (it will wrap the range if
// necessary).
type randomScanStrategy struct {
	rand *rand.Rand
}

func (rss randomScanStrategy) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
	if count >= max {
		return 0, false
	}
	offset := rss.rand.Intn(max)
	for i := 0; i < max; i++ {
		at := (offset + i) % max
		if allocated.Bit(at) == 0 {
			return at, true
		}
	}
	return 0, false
}
```

The proposal is to implement a new strategy on the Service IP allocation bitmap that avoids to use the first 32 values

> if there are free values on the upper block of the range.

Reviewer 4: That seems weird. I would expect that if there was a reserved range, that it would be reserved unconditionally, not "reserved until we run out of unreserved addresses".

Also, it's weird to hardcode the reserved range to 32 addresses regardless of the size of the service CIDR...

Reviewer: Heh.  We went back and forth on this a lot in discussing.

If your service CIDR is /27, you only get 32 service IPs (IIRC, it's 30 actually, we avoid the first and last).  At that point we can debate whether it's OK to have so few and if, at that point, we can make many guarantees.

If your CIDR is bigger than that, do we need to reserve more than 32 IPs?  If so, do we need a "cascade" of blocks?

e.g. something like `min(32, cidrSize/2)` is attractive, but only useful if we think there are a lot of users with really small ranges.

Something more dynamic like "always prefer the upper half unless it is full" is more general, I guess.  Is it worth the extra effort?

E.g. assume a /22 service range (1024 IPs).   First look at [512]-[1023].  If that's full, look at [256-511].  If that's full, look at [128-255].  It seems like that gives the caller influence where it isn't needed, which could be a vector for some sort of attack.

`min(max(16, cidrSize/16), 256)` ?  At /28 it would be 16 (full range).  At /27 - /24 it would be 16 (part of the range).  At /23 it would be 32.  /22 -> 64. /21 ->128.  /20 and bigger -> 256.

Again, worth the effort?

Author: what about making it a fixed percentage? we subdivide in two ranges, 

```

           ┌─────────────┬─────────────────────────────────────────────┐
           │             │                                             │
           │             │                                             │
           │             │                                             │
           │   static    │                    dynamic                  │
           │             │                                             │
           │             │                                             │
           │             │                                             │
           └─────────────┴─────────────────────────────────────────────┘

           ◄────────────► ◄────────────────────────────────────────────►

                15%                              85%
```

Reviewer: "15%" must be a typo, since it is not a power of two.  I think you meant to type "12.5%" :)

Seriously though, I think percents are OK for humans to get a big-picture, but harder to be precise about.

`min(max($min, cidrSize/$step), $max)` is easy to describe as "never less than $min or more than $max, with a graduated step function between them.  We can argue about values for those variables (or even tweak them over time if we get feedback), but the logic still seems appropriate?

I proposed:

$min = 16
$max = 256
$step = 16

But I am fine to haggle :)

Author: I'm fine with anything meanwhile users can understand it

Author: updated with the suggested formula

Reviewer 4: FTR I meant that 32 seemed like too many reserved IPs for small service CIDRs, not that I thought we should reserve more for large CIDRs.

But I guess it's reasonable to assume that users with larger clusters are likely to have more use cases for fixed-IP services, so I guess giving them a larger range makes sense. And if we want the reserved size to (a) always be a power of 2, and (b) always include the .10 address, then I guess 16 has to be the minimum value. And given that the "reserved" range is more like a "preferred" range, then even 16 isn't too large for a /27

```go
// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
// forward looking for the next available address. The big.Int range is subdivided so it will try
// to allocate first from the reserved block of addresses (it will wrap the reserved subrange if necessary).
// If there is no free address it will try to allocate one from the reserved block too.
type randomScanReservedStrategy struct {
	rand     *rand.Rand
	reserved int
}
```

In order to avoid any possible issue with cluster with small subnets, this strategy will only work if the service IP range

> is greater than 64, i.e., these apply only for Service subnets greater than IPv4 /26 or IPv6 /122.

Reviewer 4: If your service CIDR only has 64 addresses in it, then the odds of someone randomly getting assigned the .10 address are much higher than in the case where you have 65,536 addresses, so it seems like the reserved range is even more necessary...

Author: 64 is just is the next range available after 32, that is the number that we choose for the reserved range, so you have at least the same for static and reserved
We can choose different numbers, it just seems that 32 should be enough

Reviewer 4: No, I mean, the basic premise of this KEP is "it's important for us to avoid the 1-in-65,536 chance of a Service getting assigned the IP that CoreDNS wants in a cluster with a /16 service CIDR" so it seems weird to say that we aren't going to solve the problem in the /26 case where it is 1,000 times more likely to happen...

Author: yeah, it seems more important for the smaller ranges

  - Feature gate name: ServiceIPStaticSubrange
  - Components depending on the feature gate: kube-apiserver
- [x] Other
  - Describe the mechanism:
  - The feature require to restart the apiserver to use the new allocation strategy for Services, however,
  it will not require downtime in multi-control-plane environments.
    
###### Does enabling the feature change any default behavior?

Dynamic allocated IPs will not be chosen from the first 32 values of the Service IP range, that is transparent to the user,
since the IP is and was already random, it doesn't matter from within the range the IP is obtained.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes.

###### What happens if we reenable the feature if it was previously rolled back?

Nothing, this only affects new requests and how these new request obtain the dynamically assigned IPs.

###### Are there any tests for feature enablement/disablement?

No need to add new test, current tests should keep working as today.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

> This only affects the creation of new Services without an IP specified, there is no risk of impact on running workloads.

Reviewer 4: What happens if the cluster is already using IPs in the reserved range?
(Yes, it's fine, but you should say that explicitly.)

    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> 

Author: self reminder to add https://github.com/kubernetes/enhancements/pull/1881 as an alternative
